crates/accelerate/src/synthesis/linear_phase/cnot_phase_synth.rs

```diff
@@ -283,10 +283,7 @@ pub fn synth_cnot_phase_aam(
     }
 
     let state_bool = state.mapv(|x| x != 0);
-    let mut instrs = _synth_cnot_count_full_pmh(state_bool, section_size)
-        .into_iter()
-        .rev()
-        .collect();
+    let mut instrs = synth_pmh(state_bool, section_size).rev().collect();
```

Comment:
Is it possible to write the `instructions` as iterator? That would allow us to avoid the collections and have `CircuitData` directly consume the iterator (and likely be a bit faster 🙂)
```
let insts = instructions.chain(synth_pmh(...).rev());
CircuitData::from_standard_gates(..., insts,...);
```

Reply:
I tried writing `instrucitons` as iterator, instead of creating it as vector and pushing each value.
I did something like:
```rust
 type Instructions = (StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>);
    let mut instructions: Box<dyn Iterator<Item = Instructions>> = Box::new(iter::empty());

    // some more code

          instructions = Box::new(instructions.chain(match rust_angles.remove(index) {
                             gate if gate == "t" => iter::once((
                                  StandardGate::TGate,
                                  smallvec![],
                                  smallvec![Qubit(_ep as u32)],
                              )),
                              gate if gate == "tdg" => iter::once((
                                  StandardGate::TdgGate,
                                  smallvec![],
                                  smallvec![Qubit(_ep as u32)],
                              )),
                              gate if gate == "s" => iter::once((
                                  StandardGate::SGate,
                                  smallvec![],
                                  smallvec![Qubit(_ep as u32)],
                              )),
                              gate if gate == "sdg" => iter::once((
                                  StandardGate::SdgGate,
                                  smallvec![],
                                  smallvec![Qubit(_ep as u32)],
                              )),
                              gate if gate == "z" => iter::once((
                                  StandardGate::ZGate,
                                  smallvec![],
                                  smallvec![Qubit(_ep as u32)],
                              )),
                              angles_in_pi => iter::once((
                                  StandardGate::PhaseGate,
                                  smallvec![Param::Float(
                                      (angles_in_pi.parse::<f64>()?) % PI
                                  )],
                                  smallvec![Qubit(_ep as u32)],
                              )),
                          }));
```
I didn't want to use dynamic dispatch so tried to get rid of `Box`, but It looks like I cannot do away with 'Box' in this particular case.
Now, this made the performance drop x6 times. So, I did `git stash`

and, tried this instead:
I made the return type of `synth_pmh` as `impl DoubleEndedIterator<Item = Instructions>`
and, for `synth_cnot_phase_aam` at the very last I did `let instructions = instructions.into_iter().chain(synth_pmh(state_bool, section_size).rev());`
These changes were similar to what I did in [156da41], and, consequently it showed the same tiny performance drop, which was the reason I reverted back in [b236d62]